src/util.c

```diff
@@ -45,3 +46,40 @@ void *ecalloc(size_t nmemb, size_t size)
         err(EXIT_FAILURE, "calloc");
     return p;
 }
+
+void *erealloc(void *ptr, size_t size)
+{
+    void *p = realloc(ptr, size);
+    if (!p)
+        err(EXIT_FAILURE, "realloc");
+    return p;
+}
+
+void streamReserve(Stream *buf, size_t size)
+{
+    assert(buf->off <= buf->cap);
+    size_t avail = buf->cap - buf->off;
+    if (avail < size) {
+        buf->cap += MAX(size, 128); /* be a bit greedy when allocating */
+        buf->buf = erealloc(buf->buf, buf->cap);
+    }
+    assert((buf->off + size) <= buf->cap);
+}
+
+void streamChar(Stream *buf, char ch)
+{
+    streamReserve(buf, 1);
+    buf->buf[buf->off++] = ch;
+}
+
+void streamMem(Stream *buf, const char *str, size_t n)
+{
+    streamReserve(buf, n);
+    for (size_t i = 0; i < n; ++i)
+        buf->buf[buf->off++] = str[i];
+}
+
+void streamStr(Stream *buf, const char *str)
+{
+    streamMem(buf, str, strlen(str));
```

Comment:
This should use `strlcpy()` and reserve more memory and try again if it truncated.

Reply:
The logic that strlcpy propogates `copy => realloc => restart copy` is overcomplicated, unnatural and inefficient. "Reserve enough space and then copy over" is a significantly simpler and straightforward way to do things.

Reply:
strlcpy() will iterate once if there's enough memory, the current code always iterates twice. This at least has the advantage of being short, so either way is fine I guess.

Reply:
> the current code always iterates twice

Unless the data is bigger than the L1d cache (unlikely), memory accesses are cheap. But the main difference between copying a nul-string vs copying a fixed len buffer is that the string copy's loop condition is typically some form of `s[i] != '\0'` which means the cpu needs to constantly keep checking for this condition and cannot (*observably*) execute the next loop iteration. But the copy len being known in the buffer case allows the processor to execute the next iteration before the first one finishes (due to optimizations such as [Out of Order execution](https://en.wikipedia.org/wiki/Out-of-order_execution)).

This is of course micro-optimization that's not relavant for our purposes (*main* efficiency gain in this patch is removal of strlcat _on each byte_).